GUI: Display warning when attempting address reuse

CryptAxe opened the pull request:
Added address reuse key 'au' to destdata

Added informative QMessageBox popup warning when user attempts address reuse

CryptAxe committed:
GUI: Display warning when attempting address reuse

Added address reuse key 'au' to destdata

Added informative QMessageBox popup warning when user attempts address reuse

CryptAxe commented:
Related to https://github.com/bitcoin/bitcoin/issues/3266

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -263,6 +263,27 @@ void SendCoinsDialog::on_sendButton_clicked()
+        {
+            QString reuseWarningText = tr(
+                        "You have sent to the Bitcoin address %1 already!\n\n"
+                        "Sending to the same address twice is technically not supported and "
```
I think the part `technically not supported` is not correct.

TheBlueMatt replied:
Indeed, it is supported fine, but it is highly discouraged.

luke-jr replied:
It would not be a bug to stop scanning for transactions paying to an address once it has been used. It "works" only by accident.

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -263,6 +263,27 @@ void SendCoinsDialog::on_sendButton_clicked()
+                        "will reduce your own security as well as the privacy of all Bitcoin users."
+                    ).arg(rcp.address) + "\n";
+
+            QMessageBox reuseMessage("Warning: address reuse!",
```
I can't find this constructor in the Qt docs...

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -263,6 +263,27 @@ void SendCoinsDialog::on_sendButton_clicked()
+            QMessageBox reuseMessage("Warning: address reuse!",
+                                     reuseWarningText,
+                                     QMessageBox::Warning,
+                                     QMessageBox::Yes | QMessageBox::Escape,
```
Yes | Escape doesn't look like a normal combination?

luke-jr commented on src/qt/walletmodel.cpp:
```diff
@@ -355,6 +355,9 @@ WalletModel::SendCoinsReturn WalletModel::sendCoins(WalletModelTransaction &tran
 
                 std::map<CTxDestination, CAddressBookData>::iterator mi = wallet->mapAddressBook.find(dest);
 
+                std::string key = "au"; // "au" prefix = "address used" in destdata
```
`"used"` seems more straightforward, but more importantly, this should be moved to a `static const` in a `src/wallet/` header.

jonasschnelli commented on src/qt/walletmodel.cpp:
```diff
@@ -591,6 +594,24 @@ bool WalletModel::isSpent(const COutPoint& outpoint) const
+bool WalletModel::isUsed(const std::string& address) const
+{
+    LOCK2(cs_main, wallet->cs_wallet);
+    for (const std::pair<CTxDestination, CAddressBookData>& item : wallet->mapAddressBook)
```
Instead of looping, would a std::map::find be possible (`std::map<CTxDestination, CAddressBookData>::iterator mi = wallet->mapAddressBook.find(dest);`?

jonasschnelli commented:
General concept ACK.
Am I correct that this patch does not warn for "send-agains" to addresses sent before this patch was active?

MarcoFalke commented:
Agree that the "technically not supported" is confusing to users. We
should just mention "is (highly) discouraged".

CryptAxe committed:
Update isUsed, destdata used key, warning text

CryptAxe commented:
Okay I changed the warning message because if there is confusion even here then it needs to be more clear.  I also updated the message box code so that it isn't using an obsolete constructor and removed the extra escape.

This warning will only be displayed for addresses which are used after the software has been updated. It would be possible though for someone to add an update later that would scan through the wallet transactions and mark the destinations of those as used. Also the wallet already had a function to read the destdata by using std::map find, but it wasn't being used anywhere as far as I can tell. I replaced the loop with that.

Thanks everyone for the feedback, hopefully my update https://github.com/bitcoin/bitcoin/pull/9722/commits/8f01b1f8f6e918b0a52397e12257cdc1f47dfc9d addressed all of the comments here

CryptAxe committed:
Add destdata after updating address book

CryptAxe commented:
I am going to go through some tests now, but everything is ready to be reviewed again.

CryptAxe commented:
Alright I'm happy with the code. Anyone have a problem with the updated warning message?

jonasschnelli commented:
Code looks good to me. I dislike the `"1"`, but I guess this is just how we have to do it with the flexible per wtx map.

Though I'm unsure about the update path of this feature. If we only mark new destinations, this may result in a false awareness of address reuse because some addresses you may reuse (the old ones before the upgrade) won't show the warning and users may think it's okay to reuse them.

A solution would be, to either A) loop through the transactions that where made before the feature was active or to B) migrate at first start with the new feature (loop through all wtxs, set the `used` dest data.

I'd like to hear more opinions on that.

luke-jr commented:
I'd prefer B, but it would only allow one reuse before it'd get flagged, right?

CryptAxe commented:
Yeah the way it works right now, and with B, the code will only allow
tagging on first use. I'm fine switching to B or whatever makes it more
clear but 1 = true made sense to me.

On Feb 10, 2017 1:07 AM, "Luke Dashjr" <notifications@github.com> wrote:

> I'd prefer B, but it would only allow one reuse before it'd get flagged,
> right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9722#issuecomment-278892806>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHu1NptENp_ak43bJO0zlnipHctySWBlks5rbCjVgaJpZM4L6ce2>
> .
>

jonasschnelli commented:
Any intentions to work this towards the solution "B" (migrate at first start with the new feature (loop through all wtxs, set the used dest data.)?

CryptAxe commented:
Yes, I was just waiting for some feedback on that before I went ahead and
added the code. Since I think it's a good idea and nobody seems to take
issue I'll finish this up next time I have a break.

On Mar 17, 2017 6:47 AM, "Jonas Schnelli" <notifications@github.com> wrote:

> Any intentions to work this towards the solution "B" (migrate at first
> start with the new feature (loop through all wtxs, set the used dest data.)?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9722#issuecomment-287357677>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHu1Nm20_a6Eq_-MxgEfKp5zebvJ9MU2ks5rmo7ggaJpZM4L6ce2>
> .
>

jonasschnelli commented:
Closing due to inactivity (feel free to reopen once someone takle option B or there is consensus about doing solution A).
